adam/curriculum_to_html.py

```diff
@@ -339,6 +332,42 @@ def render_property(prop: PropertyPerception) -> str:
         )
         all_relations = first_frame_relations.union(second_frame_relations)
 
+        # Here we add the perceived objects to a NetworkX DiGraph with PART_OF relations being the
+        # edges between objects. This allows us to do pre-order traversal of the Graph to make a
+        # nested <ul></ul> for the objects rather than a flat list.
+        graph = DiGraph()
+        root = ObjectPerception("root")
+        graph.add_node(root)
+
+        for object_ in all_objects:
+            graph.add_node(object_)
+            graph.add_edge(root, object_)
+
+        for relation_ in all_relations:
+            if relation_.relation_type == PART_OF:
+                graph.add_edge(relation_.first_slot, relation_.second_slot)
+
+        # Next, we render objects, together with their properties, using preorder DFS Traversal
+        output_text.append("\t\t<h5>Perceived Objects</h5>\n\t\t")
+        visited = set()
+
+        def dfs_walk(node):
```

Comment:
Don't reinvent the wheel :-) : https://networkx.github.io/documentation/stable/reference/algorithms/generated/networkx.algorithms.traversal.depth_first_search.dfs_preorder_nodes.html#networkx.algorithms.traversal.depth_first_search.dfs_preorder_nodes

Reply:
Nevermind, @LichMaster98 correctly points out he needs this to get the nested list structure.